File: kotlin/timetravel/src/commonMain/kotlin/io/flowdux/timetravel/TimeTravelStore.kt
package io.flowdux.timetravel

import io.flowdux.Action
import io.flowdux.DefaultErrorProcessor
import io.flowdux.ErrorProcessor
import io.flowdux.Middleware
import io.flowdux.NoOpStoreLogger
import io.flowdux.Reducer
import io.flowdux.State
import io.flowdux.Store
import io.flowdux.createStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.datetime.Clock

class TimeTravelStore<S : State, A : Action> internal constructor(
    private val innerStore: Store<S, A>,
    private val maxHistorySize: Int,
    initialState: S?,
    initialHistory: List<StateSnapshot<S, A>>?,
) {
    private val mutex = Mutex()

    private val _history = mutableListOf<StateSnapshot<S, A>>()
    val history: List<StateSnapshot<S, A>> get() = _history.toList()

    private var _currentIndex = 0
    val currentIndex: Int get() = _currentIndex

    private val _state: MutableStateFlow<S>
    val state: StateFlow<S> get() = _state.asStateFlow()

    val currentState: S get() = _state.value

    val isClosed: Boolean get() = innerStore.isClosed

    val canUndo: Boolean get() = _currentIndex > 0
    val canRedo: Boolean get() = _currentIndex < _history.size - 1

    init {
        when {
            !initialHistory.isNullOrEmpty() -> {
                _history.addAll(
                    initialHistory.mapIndexed { idx, snapshot ->
                        snapshot.copy(index = idx)
                    },
                )
                _currentIndex = _history.size - 1
                _state = MutableStateFlow(_history.last().currentState)
            }
            initialState != null -> {
                _history.add(
                    StateSnapshot(
                        index = 0,
                        action = null,
                        previousState = null,
                        currentState = initialState,
                        timestamp = currentTimeMillis(),
                    ),
                )
                _state = MutableStateFlow(initialState)
            }
            else -> error("Either initialState or initialHistory must be provided")
        }
    }

    internal fun recordStateChange(action: A, previousState: S, newState: S) {
        if (_currentIndex < _history.size - 1) {
            while (_history.size > _currentIndex + 1) {
                _history.removeAt(_history.size - 1)
            }
        }

        val newIndex = _history.size
        _history.add(
            StateSnapshot(
                index = newIndex,
                action = action,
                previousState = previousState,
                currentState = newState,
                timestamp = currentTimeMillis(),
            ),
        )

        while (_history.size > maxHistorySize && _history.size > 1) {
            _history.removeAt(0)
            _history.forEachIndexed { idx, snapshot ->
                if (snapshot.index != idx) {
                    _history[idx] = snapshot.copy(index = idx)
                }
            }
        }

        _currentIndex = _history.size - 1
        _state.value = newState
    }

    fun dispatch(action: A) {
        innerStore.dispatch(action)
    }

    suspend fun undo(): Boolean = mutex.withLock {
        if (!canUndo) return false
        _currentIndex--
        _state.value = _history[_currentIndex].currentState
        true
    }

    suspend fun redo(): Boolean = mutex.withLock {
        if (!canRedo) return false
        _currentIndex++
        _state.value = _history[_currentIndex].currentState
        true
    }

    suspend fun jumpTo(index: Int): Boolean = mutex.withLock {
        if (index < 0 || index >= _history.size) return false
        _currentIndex = index
        _state.value = _history[_currentIndex].currentState
        true
    }

    suspend fun reset(): Boolean = jumpTo(0)

    suspend fun clear() = mutex.withLock {
        val currentState = _state.value
        _history.clear()
        _history.add(
            StateSnapshot(
                index = 0,
                action = null,
                previousState = null,
                currentState = currentState,
                timestamp = currentTimeMillis(),
            ),
        )
        _currentIndex = 0
    }

    fun close() {
        innerStore.close()
    }
}

fun <S : State, A : Action> createTimeTravelStore(
    initialState: S,
    reducer: Reducer<S, A>,
    middlewares: List<Middleware<S, A>> = emptyList(),
    errorProcessor: ErrorProcessor<A> = DefaultErrorProcessor(),
    maxHistorySize: Int = 100,
    scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
): TimeTravelStore<S, A> = createTimeTravelStoreInternal(
    initialState = initialState,
    initialHistory = null,
    reducer = reducer,
    middlewares = middlewares,
    errorProcessor = errorProcessor,
    maxHistorySize = maxHistorySize,
    scope = scope,
)

fun <S : State, A : Action> createTimeTravelStore(
    initialHistory: List<StateSnapshot<S, A>>,
    reducer: Reducer<S, A>,
    middlewares: List<Middleware<S, A>> = emptyList(),
    errorProcessor: ErrorProcessor<A> = DefaultErrorProcessor(),
    maxHistorySize: Int = 100,
    scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
): TimeTravelStore<S, A> {
    require(initialHistory.isNotEmpty()) { "initialHistory must not be empty" }
    return createTimeTravelStoreInternal(
        initialState = null,
        initialHistory = initialHistory,
        reducer = reducer,
        middlewares = middlewares,
        errorProcessor = errorProcessor,
        maxHistorySize = maxHistorySize,
        scope = scope,
    )
}

private fun <S : State, A : Action> createTimeTravelStoreInternal(
    initialState: S?,
    initialHistory: List<StateSnapshot<S, A>>?,
    reducer: Reducer<S, A>,
    middlewares: List<Middleware<S, A>>,
    errorProcessor: ErrorProcessor<A>,
    maxHistorySize: Int,
    scope: CoroutineScope,
): TimeTravelStore<S, A> {
    lateinit var timeTravelStore: TimeTravelStore<S, A>

    val timeTravelReducer =
        Reducer<S, A> { _, action ->
            reducer.reduce(timeTravelStore.currentState, action)
        }

    val historyLogger =
        object : NoOpStoreLogger<S, A>() {
            override fun onStateReduced(action: A, previousState: S, newState: S) {
                timeTravelStore.recordStateChange(action, timeTravelStore.currentState, newState)
            }
        }

    val effectiveInitialState = initialHistory?.lastOrNull()?.currentState ?: initialState!!

    val innerStore =
        createStore(
            initialState = effectiveInitialState,
            reducer = timeTravelReducer,
            middlewares = middlewares,
            errorProcessor = errorProcessor,
            logger = historyLogger,
            scope = scope,
        )

    timeTravelStore =
        TimeTravelStore(
            innerStore = innerStore,
            maxHistorySize = maxHistorySize,
            initialState = initialState,
            initialHistory = initialHistory,
        )

    return timeTravelStore
}

private fun currentTimeMillis(): Long = Clock.System.now().toEpochMilliseconds()

File: dart/flowdux/lib/src/timetravel/time_travel_store.dart
import 'package:rxdart/rxdart.dart';

import '../action.dart';
import '../error_processor.dart';
import '../middleware.dart';
import '../reducer.dart';
import '../store.dart';
import '../store_logger.dart';
import '../util/async_lock.dart';
import 'state_snapshot.dart';

/// A store wrapper that provides time travel debugging capabilities.
///
/// TimeTravelStore wraps a regular [Store] and tracks all state changes
/// in a history list, enabling undo, redo, and jump-to operations.
///
/// Example:
/// ```dart
/// final store = createTimeTravelStore<CounterState, CounterAction>(
///   initialState: CounterState(0),
///   reducer: counterReducer,
/// );
///
/// store.dispatch(IncrementAction());
/// store.dispatch(IncrementAction());
/// await store.undo(); // Back to count: 1
/// await store.redo(); // Forward to count: 2
/// await store.jumpTo(0); // Back to initial state
/// ```
class TimeTravelStore<S, A extends Action> {
  final Store<S, A> _innerStore;

  /// Maximum number of history entries to keep.
  final int maxHistorySize;

  final AsyncLock _lock = AsyncLock();
  final List<StateSnapshot<S, A>> _history = [];
  int _currentIndex = 0;
  final BehaviorSubject<S> _stateSubject;

  TimeTravelStore._({
    required Store<S, A> innerStore,
    required this.maxHistorySize,
    required S seedState,
    List<StateSnapshot<S, A>>? initialHistory,
  })  : _innerStore = innerStore,
        _stateSubject = BehaviorSubject.seeded(seedState) {
    if (initialHistory != null && initialHistory.isNotEmpty) {
      for (var i = 0; i < initialHistory.length; i++) {
        _history.add(initialHistory[i].copyWith(index: i));
      }
      _currentIndex = _history.length - 1;
    } else {
      _history.add(StateSnapshot(
        index: 0,
        currentState: seedState,
        timestamp: DateTime.now(),
      ));
    }
  }

  /// Stream of state changes.
  ///
  /// Emits the current state immediately upon subscription,
  /// then emits new states whenever the state changes.
  Stream<S> get state => _stateSubject.stream;

  /// Current state value (synchronous access).
  S get currentState => _stateSubject.value;

  /// A copy of the current history list.
  List<StateSnapshot<S, A>> get history => List.unmodifiable(_history);

  /// Current position in the history.
  int get currentIndex => _currentIndex;

  /// Whether undo is available (not at the beginning of history).
  bool get canUndo => _currentIndex > 0;

  /// Whether redo is available (not at the end of history).
  bool get canRedo => _currentIndex < _history.length - 1;

  /// Whether the store has been closed.
  bool get isClosed => _innerStore.isClosed;

  /// Records a state change in the history.
  void _recordStateChange(A action, S previousState, S newState) {
    // Truncate future history if we're not at the end
    if (_currentIndex < _history.length - 1) {
      while (_history.length > _currentIndex + 1) {
        _history.removeLast();
      }
    }

    final newIndex = _history.length;
    _history.add(StateSnapshot(
      index: newIndex,
      action: action,
      previousState: previousState,
      currentState: newState,
      timestamp: DateTime.now(),
    ));

    // Enforce max history size
    while (_history.length > maxHistorySize && _history.length > 1) {
      _history.removeAt(0);
      for (var i = 0; i < _history.length; i++) {
        if (_history[i].index != i) {
          _history[i] = _history[i].copyWith(index: i);
        }
      }
    }

    _currentIndex = _history.length - 1;
    if (newState != _stateSubject.value) {
      _stateSubject.add(newState);
    }
  }

  /// Dispatches an action to the underlying store.
  void dispatch(A action) {
    _innerStore.dispatch(action);
  }

  /// Moves to the previous state in history.
  ///
  /// Returns `true` if undo was successful, `false` if already at the beginning.
  Future<bool> undo() => _lock.synchronized(() async {
        if (!canUndo) return false;
        _currentIndex--;
        final newState = _history[_currentIndex].currentState;
        if (newState != _stateSubject.value) {
          _stateSubject.add(newState);
        }
        return true;
      });

  /// Moves to the next state in history.
  ///
  /// Returns `true` if redo was successful, `false` if already at the end.
  Future<bool> redo() => _lock.synchronized(() async {
        if (!canRedo) return false;
        _currentIndex++;
        final newState = _history[_currentIndex].currentState;
        if (newState != _stateSubject.value) {
          _stateSubject.add(newState);
        }
        return true;
      });

  /// Jumps to a specific index in the history.
  ///
  /// Returns `true` if jump was successful, `false` if the index is out of bounds.
  Future<bool> jumpTo(int index) => _lock.synchronized(() async {
        if (index < 0 || index >= _history.length) return false;
        _currentIndex = index;
        final newState = _history[_currentIndex].currentState;
        if (newState != _stateSubject.value) {
          _stateSubject.add(newState);
        }
        return true;
      });

  /// Resets to the initial state (index 0).
  ///
  /// History is preserved. Returns `true` if successful.
  Future<bool> reset() => jumpTo(0);

  /// Clears all history and keeps only the current state as the new initial state.
  Future<void> clear() => _lock.synchronized(() async {
        final current = _stateSubject.value;
        _history.clear();
        _history.add(StateSnapshot(
          index: 0,
          currentState: current,
          timestamp: DateTime.now(),
        ));
        _currentIndex = 0;
      });

  /// Closes the underlying store and releases resources.
  ///
  /// This method is idempotent - calling it multiple times has no effect.
  Future<void> close() async {
    await _innerStore.close();
    if (!_stateSubject.isClosed) {
      await _stateSubject.close();
    }
  }
}

/// Creates a new [TimeTravelStore] with the given configuration.
///
/// Example:
/// ```dart
/// final store = createTimeTravelStore<CounterState, CounterAction>(
///   initialState: CounterState(0),
///   reducer: counterReducer,
///   maxHistorySize: 50,
/// );
/// ```
TimeTravelStore<S, A> createTimeTravelStore<S, A extends Action>({
  required S initialState,
  required Reducer<S, A> reducer,
  List<Middleware<S, A>> middlewares = const [],
  ErrorProcessor<A>? errorProcessor,
  int maxHistorySize = 100,
}) {
  return _createTimeTravelStoreInternal(
    initialState: initialState,
    initialHistory: null,
    reducer: reducer,
    middlewares: middlewares,
    errorProcessor: errorProcessor,
    maxHistorySize: maxHistorySize,
  );
}

/// Creates a [TimeTravelStore] from a previously saved history.
///
/// Throws [ArgumentError] if [initialHistory] is empty.
///
/// Example:
/// ```dart
/// final store = createTimeTravelStoreFromHistory<CounterState, CounterAction>(
///   initialHistory: savedHistory,
///   reducer: counterReducer,
/// );
/// ```
TimeTravelStore<S, A> createTimeTravelStoreFromHistory<S, A extends Action>({
  required List<StateSnapshot<S, A>> initialHistory,
  required Reducer<S, A> reducer,
  List<Middleware<S, A>> middlewares = const [],
  ErrorProcessor<A>? errorProcessor,
  int maxHistorySize = 100,
}) {
  if (initialHistory.isEmpty) {
    throw ArgumentError('initialHistory must not be empty');
  }
  return _createTimeTravelStoreInternal(
    initialState: null,
    initialHistory: initialHistory,
    reducer: reducer,
    middlewares: middlewares,
    errorProcessor: errorProcessor,
    maxHistorySize: maxHistorySize,
  );
}

TimeTravelStore<S, A> _createTimeTravelStoreInternal<S, A extends Action>({
  S? initialState,
  List<StateSnapshot<S, A>>? initialHistory,
  required Reducer<S, A> reducer,
  required List<Middleware<S, A>> middlewares,
  ErrorProcessor<A>? errorProcessor,
  required int maxHistorySize,
}) {
  late TimeTravelStore<S, A> timeTravelStore;

  S timeTravelReducer(S _, A action) {
    return reducer(timeTravelStore.currentState, action);
  }

  final historyLogger = _TimeTravelLogger<S, A>(
    (action, previousState, newState) {
      timeTravelStore._recordStateChange(
        action,
        timeTravelStore.currentState,
        newState,
      );
    },
  );

  final effectiveInitialState =
      initialHistory != null && initialHistory.isNotEmpty
          ? initialHistory.last.currentState
          : initialState!;

  final innerStore = createStore<S, A>(
    initialState: effectiveInitialState,
    reducer: timeTravelReducer,
    middlewares: middlewares,
    errorProcessor: errorProcessor,
    logger: historyLogger,
  );

  timeTravelStore = TimeTravelStore<S, A>._(
    innerStore: innerStore,
    maxHistorySize: maxHistorySize,
    seedState: effectiveInitialState,
    initialHistory: initialHistory,
  );

  return timeTravelStore;
}

/// Internal logger that records state changes for time travel.
///
/// Implements [StoreLogger] directly (not extending [NoOpStoreLogger])
/// so that the Store's logging pipeline is active and [onStateReduced]
/// is called for every state transition.
class _TimeTravelLogger<S, A extends Action> implements StoreLogger<S, A> {
  final void Function(A action, S previousState, S newState) _onStateReduced;

  _TimeTravelLogger(this._onStateReduced);

  @override
  void onStateReduced(A action, S previousState, S newState) {
    _onStateReduced(action, previousState, newState);
  }

  @override
  void onActionDispatched(A action) {}

  @override
  void onMiddlewareProcessing(String middlewareName, A action) {}

  @override
  void onMiddlewaresCompleted(A action) {}

  @override
  void onFlowHolderActionEmitted(A action) {}

  @override
  void onErrorOccurred(Object error, StackTrace stackTrace) {}

  @override
  void onErrorHandled(A action) {}

  @override
  void onDispatchAfterClose(A action) {}
}
